### ISSUE TITLE
Computation graphs overcomplicated when reading/writing  partitioned parquet files?

### ISSUE THREAD

**Issue by andrethrill** [dask/dask#3831]
# TLDR

I'm having an issue where I don't understand why some Dask Computation graphs are more complicated than I would expect for some simple computations.

Below follows full code for reproducibility.

The problem itself shows in the section `The Problem`.

Something that I think is also related to the problem (Incoherent divisions) is shown in the end of the section `The 4 parquet partitions individually`/` This may be part of the problem:`.

Questions start with `Question (n):` along the text when `n` is the question number. I copy them here:

- **Question 1:** When checking the divisions, notice how the last division is wrong. It coincides with the end of the FULL dataframe `df`, although `df_a`'s index only goes until `2015-01-02 17:39:00`. I suspect this incoherency may be the cause of what I'm going to show next. Is this last division expected to be as it shows below?
- **Question 2:** Shouldn't Dask be able to generate a simpler computation graph?
- **Question 3:** Similarly to above, shouldn't Dask be able to generate a simpler computation graph?
- **Question 4:** Given that the dataframe is being read from a parquet file that is partitioned on `categoricals`, shouldn't the `.loc` be smart to realize 'a' can only be in 4 of the 16 Row Groups?

# Imports

```python
import numpy as np
import pandas as pd
import dask.dataframe as dd
```

```python
from fastparquet import ParquetFile
```

# Generate dummy data and save to a partitioned parquet file

```python
def generate_df(size):
    df = pd.Series(np.arange(0, size), name='unique').to_frame()

    df['dates'] = pd.date_range('2015-01-01', periods=size, freq='1min', name='dates')
    df['ints'] = np.linspace(0, 100, size, endpoint=False, dtype=int)
    df['floatbools'] = np.append(np.zeros(int(size / 2)), np.ones(int(size / 2)))
    df['floats'] = df['ints'].astype(float)
    df['constants'] = 42
    df['uniques'] = df['unique'].astype(str)

    # casting floatbool with float dtype to boolbool with bool dtype
    df['boolbools'] = df['floatbools'].astype(bool)

    # nanints is ints with first 10% entries set to nan
    df['nanints'] = df['ints']
    df.loc[df.index < int(size / 10), 'nanints'] = np.nan

    # generate categorical column with 4 distinct values
    categoricals = []
    for value in ['a', 'b', 'c', 'd']:
        categoricals += [value] * int(size / 4)
    df['categoricals'] = categoricals
    df['nancategoricals'] = df['categoricals'].replace('d', np.nan)

    return df
```

```python
df = generate_df(10000)
```

```python
df.head()
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>dates</th>
      <th>ints</th>
      <th>floatbools</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>boolbools</th>
      <th>nanints</th>
      <th>categoricals</th>
      <th>nancategoricals</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0</td>
      <td>2015-01-01 00:00:00</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>0</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>1</th>
      <td>1</td>
      <td>2015-01-01 00:01:00</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>1</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2</th>
      <td>2</td>
      <td>2015-01-01 00:02:00</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>2</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>2015-01-01 00:03:00</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>3</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>4</th>
      <td>4</td>
      <td>2015-01-01 00:04:00</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>4</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.tail()
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>dates</th>
      <th>ints</th>
      <th>floatbools</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>boolbools</th>
      <th>nanints</th>
      <th>categoricals</th>
      <th>nancategoricals</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>9995</th>
      <td>9995</td>
      <td>2015-01-07 22:35:00</td>
      <td>99</td>
      <td>1.0</td>
      <td>99.0</td>
      <td>42</td>
      <td>9995</td>
      <td>True</td>
      <td>99.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9996</th>
      <td>9996</td>
      <td>2015-01-07 22:36:00</td>
      <td>99</td>
      <td>1.0</td>
      <td>99.0</td>
      <td>42</td>
      <td>9996</td>
      <td>True</td>
      <td>99.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9997</th>
      <td>9997</td>
      <td>2015-01-07 22:37:00</td>
      <td>99</td>
      <td>1.0</td>
      <td>99.0</td>
      <td>42</td>
      <td>9997</td>
      <td>True</td>
      <td>99.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9998</th>
      <td>9998</td>
      <td>2015-01-07 22:38:00</td>
      <td>99</td>
      <td>1.0</td>
      <td>99.0</td>
      <td>42</td>
      <td>9998</td>
      <td>True</td>
      <td>99.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>9999</th>
      <td>9999</td>
      <td>2015-01-07 22:39:00</td>
      <td>99</td>
      <td>1.0</td>
      <td>99.0</td>
      <td>42</td>
      <td>9999</td>
      <td>True</td>
      <td>99.0</td>
      <td>d</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>
</div>

```python
df = df.set_index('dates')
df = dd.from_pandas(df, npartitions=16)
```

```python
# The divisions
df.divisions
```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 10:25:00'),
     Timestamp('2015-01-01 20:50:00'),
     Timestamp('2015-01-02 07:15:00'),
     Timestamp('2015-01-02 17:40:00'),
     Timestamp('2015-01-03 04:05:00'),
     Timestamp('2015-01-03 14:30:00'),
     Timestamp('2015-01-04 00:55:00'),
     Timestamp('2015-01-04 11:20:00'),
     Timestamp('2015-01-04 21:45:00'),
     Timestamp('2015-01-05 08:10:00'),
     Timestamp('2015-01-05 18:35:00'),
     Timestamp('2015-01-06 05:00:00'),
     Timestamp('2015-01-06 15:25:00'),
     Timestamp('2015-01-07 01:50:00'),
     Timestamp('2015-01-07 12:15:00'),
     Timestamp('2015-01-07 22:39:00'))

```python
# 16 partitions as expected
print('partitions: ' + str(df.npartitions))
```

    partitions: 16

```python
df.to_parquet('./data_new', compression='snappy', engine='fastparquet', partition_on=['categoricals'])
```

# Read data

```python
pq_f = ParquetFile('./data_new')
```

## The full dataframe

```python
df = dd.read_parquet(pq_f)
```

```python
df.divisions
```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 10:25:00'),
     Timestamp('2015-01-01 20:50:00'),
     Timestamp('2015-01-02 07:15:00'),
     Timestamp('2015-01-02 17:40:00'),
     Timestamp('2015-01-03 04:05:00'),
     Timestamp('2015-01-03 14:30:00'),
     Timestamp('2015-01-04 00:55:00'),
     Timestamp('2015-01-04 11:20:00'),
     Timestamp('2015-01-04 21:45:00'),
     Timestamp('2015-01-05 08:10:00'),
     Timestamp('2015-01-05 18:35:00'),
     Timestamp('2015-01-06 05:00:00'),
     Timestamp('2015-01-06 15:25:00'),
     Timestamp('2015-01-07 01:50:00'),
     Timestamp('2015-01-07 12:15:00'),
     Timestamp('2015-01-07 22:39:00'))

```python
df.npartitions
```

    16

```python
len(df)
```

    10000

## The 4 parquet partitions individually

Let's now get all the 4 parts of the full df by reading each of them into a different variable:

```python
df_a = dd.read_parquet(pq_f, filters=[('categoricals', '==', 'a')])
df_b = dd.read_parquet(pq_f, filters=[('categoricals', '==', 'b')])
df_c = dd.read_parquet(pq_f, filters=[('categoricals', '==', 'c')])
df_d = dd.read_parquet(pq_f, filters=[('categoricals', '==', 'd')])
```

```python
df_a.tail()
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>ints</th>
      <th>floatbools</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>boolbools</th>
      <th>nanints</th>
      <th>nancategoricals</th>
      <th>categoricals</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2015-01-02 17:35:00</th>
      <td>2495</td>
      <td>24</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>42</td>
      <td>2495</td>
      <td>False</td>
      <td>24.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-02 17:36:00</th>
      <td>2496</td>
      <td>24</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>42</td>
      <td>2496</td>
      <td>False</td>
      <td>24.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-02 17:37:00</th>
      <td>2497</td>
      <td>24</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>42</td>
      <td>2497</td>
      <td>False</td>
      <td>24.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-02 17:38:00</th>
      <td>2498</td>
      <td>24</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>42</td>
      <td>2498</td>
      <td>False</td>
      <td>24.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-02 17:39:00</th>
      <td>2499</td>
      <td>24</td>
      <td>0.0</td>
      <td>24.0</td>
      <td>42</td>
      <td>2499</td>
      <td>False</td>
      <td>24.0</td>
      <td>a</td>
      <td>a</td>
    </tr>
  </tbody>
</table>
</div>

Here we can see `df_a` corresponds to only 4 Parquet Row groups (or 4 Dask partitions) as expected:

```python
df_a.visualize()
```

![output_26_0](https://user-images.githubusercontent.com/25300892/43454725-d573d282-94b5-11e8-9e37-1cc0f82d9e0e.png)

### This may be part of the problem:

**Question 1:** When checking the divisions, notice how the last division is wrong. It coincides with the end of the FULL dataframe `df`, although `df_a`'s index only goes until `2015-01-02 17:39:00`. I suspect this incoherency may be the cause of what I'm going to show next. Is this last division expected to be as it shows below?

```python
df_a.divisions
```

    (Timestamp('2015-01-01 00:00:00'),
     Timestamp('2015-01-01 10:25:00'),
     Timestamp('2015-01-01 20:50:00'),
     Timestamp('2015-01-02 07:15:00'),
     Timestamp('2015-01-07 22:39:00'))

```python
len(df_a)
```

    2500

## Quick test checking dask is smart when `loc` ing the index:

```python
df.loc['2015-01-01 00:00:00':'2015-01-01 10:24:00'].visualize()
```

![output_32_0](https://user-images.githubusercontent.com/25300892/43454763-ecb9e31e-94b5-11e8-8dcb-5b0456a0ce2d.png)

Cool! Dask is smart when doing the loc because it is the index and its know the partitions

# The problem(s)

## Case 1

Imagine that, from the full dataframe `df`, I want to create a new column with new values but only for the rows where column `categoricals=='a'`:

```python
dicts={}

dicts['new_col'] = df_a['ints'] - 1

df_new = df.assign(**dicts)
```

I would expect the computation graph of the above code to be fairly simple. Basically, it needs to:
- Read the 4 Parquet Row Groups that correspond `df_a`
- Take the columns `ints` and subtract 1
- Merge with `df`. No repartitioning should be required because both `df` and `df_a` have aligned partitions (except for the bogus last `df_a` partition I highlighted above) 

Unfortunately, if we visualize the computation graph below we can see that, indeed, the 4 `df_a` Parquet Row Groups are being read. But, we can see a log of branches coming from the last `df_a` Row Group and going to different other row groups. Also, we can a lot of  `repartition-merge` block which I think are unnecessary:

```python
df_new.visualize()
```

![output_39_0](https://user-images.githubusercontent.com/25300892/43454780-f4c042c4-94b5-11e8-877f-0887541c9a8f.png)

**Question 2:** Shouldn't Dask be able to generate a simpler computation graph?

## Case 2

Also tried using `merge` but the result seems to be the same:

```python
df_temp = df_a['ints'] - 1
```

It does accept to merge things to the original and larger dataframe but it has a lot of repartitions!

```python
df_n = df.merge(df_temp.to_frame(), how='left', left_index=True, right_index=True)
```

```python
df_n.visualize()
```

![output_46_0](https://user-images.githubusercontent.com/25300892/43454799-fba04cf6-94b5-11e8-8ee1-9e9706252718.png)

**Question 3:** Similarly to above, shouldn't Dask be able to generate a simpler computation graph?

# Case 4

```python
df = dd.read_parquet(pq_f)
```

```python
df.head()
```

<div>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>unique</th>
      <th>ints</th>
      <th>floatbools</th>
      <th>floats</th>
      <th>constants</th>
      <th>uniques</th>
      <th>boolbools</th>
      <th>nanints</th>
      <th>nancategoricals</th>
      <th>categoricals</th>
    </tr>
    <tr>
      <th>dates</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2015-01-01 00:00:00</th>
      <td>0</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>0</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:01:00</th>
      <td>1</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>1</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:02:00</th>
      <td>2</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>2</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:03:00</th>
      <td>3</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>3</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
    <tr>
      <th>2015-01-01 00:04:00</th>
      <td>4</td>
      <td>0</td>
      <td>0.0</td>
      <td>0.0</td>
      <td>42</td>
      <td>4</td>
      <td>False</td>
      <td>NaN</td>
      <td>a</td>
      <td>a</td>
    </tr>
  </tbody>
</table>
</div>

```python
df.loc[df['categoricals'] == 'a'].visualize()
```

![output_51_0](https://user-images.githubusercontent.com/25300892/43454807-02a8f480-94b6-11e8-9938-dd90081a4be3.png)

**Question 4:** Given that the dataframe is being read from a parquet file that is partitioned on `categoricals`, shouldn't the `.loc` be smart to realize 'a' can only be in 4 of the 16 Row Groups?

**Comment by martindurant** [comment]
Agree that filtering should have the upper bound on the last division set to the max of the last row-group that fits. This seems to be a bug in [this line](https://github.com/dask/dask/blob/master/dask/dataframe/io/parquet.py#L293), which explicitly references the last division. It would be interesting to see that, if you set the last divion to the correct value, whether this simplifies the graph as you hope.

> shouldn't the .loc be smart to realize 'a' can only be in 4 of the 16 Row Groups

probably no - only the min/max of the index column is converted into divisions, but conditions on other columns are not pushed down. This is the reason that the filters keyword still exists.

**Comment by andrethrill** [comment]
I finally had the time to look further into this:


As per your suggestion, @martindurant , I've replaced the code:

```python
        if index_name in minmax:
            divisions = (list(minmax[index_name]['min']) +
                         [minmax[index_name]['max'][-1]])
            divisions = [divisions[i] for i, rg in enumerate(pf.row_groups)
                         if rg in rgs] + [divisions[-1]]
```

by:

```python
        if index_name in minmax:
            divisions = minmax[index_name]
            idx_lst = [i for i, rg in enumerate(pf.row_groups)
                         if rg in rgs]
            divisions = [divisions['min'][i]
                         for i in idx_lst] + [divisions['max'][idx_lst[-1]]]
```

And this does fix the divisions issue! Which is great:
```python
df_a.divisions
```
```
(Timestamp('2015-01-01 00:00:00'),
 Timestamp('2015-01-01 10:25:00'),
 Timestamp('2015-01-01 20:50:00'),
 Timestamp('2015-01-02 07:15:00'),
 Timestamp('2015-01-02 17:39:00'))
```

Unfortunately it does not simplify the computation graph. It's the same as before.

Still, I've kept doing some more tests and I start to believe the reason is not the partitioned parquet file:


### Loc is efficient but assign does not seem to be

If I generate the same `df` as in the beginning of the issue description but save it without `partitions_on`:

`df.to_parquet('./data_new', compression='snappy', engine='fastparquet')`

and then do:

```python
df = dd.read_parquet('./data_new')


dicts={}
dicts['new_col'] = df.loc['2015-01-01 00:00:00':'2015-01-01 10:22:00', 'ints'] - 1
df_new = df.assign(**dicts)

df_new.visualize()
```
![unknown](https://user-images.githubusercontent.com/25300892/44414360-7bacc400-a565-11e8-905e-c1c2608ac07c.png)

For reference, the divisions are:

```python
df.divisions
```
```
(Timestamp('2015-01-01 00:00:00'),
 Timestamp('2015-01-01 10:25:00'),
 Timestamp('2015-01-01 20:50:00'),
 Timestamp('2015-01-02 07:15:00'),
 Timestamp('2015-01-02 17:40:00'),
 Timestamp('2015-01-03 04:05:00'),
 Timestamp('2015-01-03 14:30:00'),
 Timestamp('2015-01-04 00:55:00'),
 Timestamp('2015-01-04 11:20:00'),
 Timestamp('2015-01-04 21:45:00'),
 Timestamp('2015-01-05 08:10:00'),
 Timestamp('2015-01-05 18:35:00'),
 Timestamp('2015-01-06 05:00:00'),
 Timestamp('2015-01-06 15:25:00'),
 Timestamp('2015-01-07 01:50:00'),
 Timestamp('2015-01-07 12:15:00'),
 Timestamp('2015-01-07 22:39:00'))
```

- I've shown before that the computation graph for `df.loc['2015-01-01 00:00:00':'2015-01-01 10:22:00', 'ints']` is efficient, but when I do `assign`, I don't understand why are all the partitions being "touched" if I'm assigning a new column with date ranges only within the first partition? 
- Even if all the partitions have to be "touched" because we have to put NaNs in all the remainings partitions for the new columns, why are all the `repartion-split` and `repartition-merge` in there?

**Comment by quasiben** [comment]
I believe https://github.com/dask/dask/pull/3931 resolved this issue.  @martindurant if that is correct, I am happy to close, or you can

**Comment by martindurant** [comment]
Yes, closing, although there are certainly parquet optimisations that we are still thinking about.
